multiple fixes to properly support multi-node scale down operations

pgoron opened the pull request:
* reverted initial attempts to fix scale down to ease integration of upstream fixes
* integrated two fixes from opensearch-k8s-operator v2..5.0
* fix detection of relocating shards
* delay re-enqueue when all conditions for scale down aren't met to increase chances to have a consistent cache
* fix cluster passing red during multi-node scale down
* prevent scaler to initiate downscale if cluster is not green

The last 4 commits are candidate for upstream contribution, but we need to catch-up with upstream first and unit tests coverage will have to be increased.

saketmht committed:
Remove node exclusion

Signed-off-by: saketmht <saket.m.mahto@oracle.com>

saketmht committed:
Fix typo

Signed-off-by: saketmht <saket.m.mahto@oracle.com>

Commit:
Revert "scaler: Fix clusterGreen condition"

This reverts commit df107a4e9e5da0d7dc2b3ab266d021d9d4245645.

Commit:
Revert "smartScaler: Check for green status before drain"

This reverts commit 76bebc81cb5feb34de43d8c4a9f342cd429d696d.

Commit:
criteo: fix relocating shard detection

operator doesn't properly detect shards being relocated and cause
nodes to be partialy drained during scale down operations.

reason is that opensearch _cat/shards api is using a slightly different
output format for relocating shards.
See https://github.com/opensearch-project/OpenSearch/blob/2.15.0/server/src/main/java/org/opensearch/rest/action/cat/RestShardsAction.java#L334

Commit:
criteo: enforce minimum delay between each scaler reconcile loop

scaler regularly re-enqueue reconciliation immediately after updating
status of opensearchcluster object.  Half of the time, next reconciliation
is done on an outdated status.

as scaler is using state machine to track progress of scale operations,
it seems better to let local cache be refreshed to increase consistency.

Commit:
criteo: fix red cluster during scale-down

* wait for nodes to have really left cluster before removing them
from routing allocation exclusion list
* when scaling down multiple nodes at once, make sure to reset
state machine to Running state between each decrease operation

this commit is candidate for upstream contribution (after adding
unit tests)

Commit:
criteo: avoid engaging scale down if cluster is not green

candidate for upstream contribution

damsallem commented on opensearch-operator/controllers/cluster_test.go:
```diff
@@ -379,7 +379,7 @@ var _ = Describe("Cluster Reconciler", func() {
 			Eventually(func() bool {
 				stsList := &appsv1.StatefulSetList{}
-				err := k8sClient.List(context.Background(), stsList, client.InNamespace(OpensearchCluster.Namespace))
+				err := k8sClient.List(context.Background(), stsList, client.InNamespace(OpensearchCluster.Name))
```
I don't expect to use the OpensearchCluster.Name here, except if the OpensearchCluster in the unit test is badly crafted and there are no Namespace or is the List of STS is mocked to answer the OpensearchCluster.Name.

pgoron replied:
Thanks David, side effect of revert of your first commits, and not fixed upstream. I will insert a commit in PR to not forget it for upstream contribution.

damsallem reviewed (approved):
One comment regarding UTs, nothing blocker.
Sounds good to me ! and in my point of view you covered it all

Commit:
criteo: fix wrong namespace passed in unit test

after deleting a node pool, unit test should look for remaining sts
into cluster namespace.
